packages/react-dom/src/client/DOMPropertyOperations.js

```diff
@@ -139,15 +141,49 @@ export function setValueForProperty(
   value: mixed,
   isCustomComponentTag: boolean,
 ) {
-  const propertyInfo = getPropertyInfo(name);
+  let propertyInfo = getPropertyInfo(name);
   if (shouldIgnoreAttribute(name, propertyInfo, isCustomComponentTag)) {
     return;
   }
+
+  if (enableCustomElementPropertySupport && isCustomComponentTag && name[0] === 'o' && name[1] === 'n') {
+    let eventName = name.replace(/Capture$/, '');
+    const useCapture = name !== eventName;
+    const nameLower = eventName.toLowerCase();
+    if (nameLower in node) {
+      eventName = nameLower;
+    }
+    eventName = eventName.slice(2);
+
+    const listenersObjName = eventName + (useCapture ? 'true' : 'false');
+    const alreadyHadListener = (node: any)._listeners && (node: any)._listeners[listenersObjName];
+
+    if (typeof value === 'function' || alreadyHadListener) {
+      if (!(node: any)._listeners) {
```

Comment:
What if a web-component also uses `_listeners` for its own use-cases?

Reply:
Great question! This happens to be what preact does in this case, and I *figured* that'd be fine.
Dan and Sebastian, do you have any ideas of where else we could store data for a node in React?

Reply:
It probably should be adding an abstraction here: https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/ReactDOMComponentTree.js#L39-L44

That uses a random suffix to avoid collisions between multiple Reacts.

I do wonder if it's worth while using a proxy or if we can just pass through. If we can add the raw listener, then we can remove it and checking whether it already exists using `ReactDOMComponentTree.getFiberCurrentPropsFromNode(...)`.

I get that it might be thrashy perf wise if it's always a new listener but a memoized one wouldn't need to.

Reply:
Thanks for the pointer to those react internals, that looks much better!

> I do wonder if it's worth while using a proxy or if we can just pass through. If we can add the raw listener, then we can remove it and checking whether it already exists using ReactDOMComponentTree.getFiberCurrentPropsFromNode(...).

I see that getFiberCurrentPropsFromNode looks at `__reactProps` on the node. Are you saying that the event listener function is already stored there somehow after we call addEventListener? Or that we should find a place in there to store the event listener?

If I remember correctly, the purpose of the proxy is to be able to quickly re-assign the actual listener function without needing to call addEventListener again because apparently it's faster.

Reply:
For a start, can we just fix this to use a unique prefix with the technique above? And then revisit.